massa-async-pool/src/pool.rs

```diff
@@ -0,0 +1,178 @@
+//! Copyright (c) 2022 MASSA LABS <info@massa.net>
+
+//! This file defines a finite size final pool of async messages for use in the context of autonomous smart contracts
+
+use std::collections::BTreeMap;
+
+use massa_models::Slot;
+
+use crate::{
+    bootstrap::AsyncPoolBootstrap,
+    changes::{AddOrDelete, AsyncPoolChanges},
+    config::AsyncPoolConfig,
+    message::{AsyncMessage, AsyncMessageId},
+};
+
+/// Represents a pool of deterministically sorted messages.
+/// The final async pool is attached to the output of the latest final slot within the context of massa-final-state.
+/// Nodes must bootstrap the final message pool when they join the network.
+#[derive(Debug, Clone)]
+pub struct AsyncPool {
+    /// async pool config
+    config: AsyncPoolConfig,
+
+    /// messages sorted by increasing ID (increasing priority)
+    messages: BTreeMap<AsyncMessageId, AsyncMessage>,
+}
+
+impl AsyncPool {
+    /// Creates an empty AsyncPool
+    pub fn new(config: AsyncPoolConfig) -> AsyncPool {
```

Comment:
I'd rather have new take an optional bootstrap snapshot for consistency with other crates. (or open an issue to change that out there)

Reply:
I don't see any options in new.
In final state : https://github.com/massalabs/massa/blob/main/massa-final-state/src/final_state.rs#L34
In ledger : https://github.com/massalabs/massa/blob/main/massa-ledger/src/ledger.rs#L77 for examples

Reply:
https://github.com/massalabs/massa/blob/932870f1df459ebb0e84a42425df5d10249aef6c/massa-graph/src/block_graph.rs#L428
https://github.com/massalabs/massa/blob/932870f1df459ebb0e84a42425df5d10249aef6c/massa-graph/src/ledger.rs#L154

Reply:
It's only in block_graph and all the rest of the code is matching the pattern we current have. I don't think we should change it.

Reply:
https://github.com/massalabs/massa/blob/932870f1df459ebb0e84a42425df5d10249aef6c/massa-proof-of-stake-exports/src/proof_of_stake.rs#L42-L46

The issue may be that we never took the time to agree on a common ground, and as the team is kinda growing, now could be the time

In my opinion, we can't really expect even specifying streaming bootstrap before may, as there is too much huge stuff going on that need to be stabilized. I hope we reach a consensus before \o/

Reply:
The downside of having the boot_pos is that we always end up with a big "if Some(xx) { [ a lot of code ] } else { [ a lot of code ] }" inside the "new" function. Having a new(config) function to create an empty structure and a from_bootstrap function to create from a bootstrap structure should be the preferred choice as it is more explicit and the functions are better split. That's why I did it in the execution module.